File: Java/3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        
        Arrays.sort(nums);
        
        int closest = Integer.MAX_VALUE;
        int closestSum = 0;
        for(int i=0; i<nums.length; i++){
            for(int j=i+1, k=nums.length-1; j<k; ){
                int sum = (nums[i] + nums[j] + nums[k]);
                
                
                int diff = sum-target;
                if(diff>0){
                    k--;
                }else{
                    diff = -diff;
                    j++;
                }
                
                if(diff < closest){
                    closest = diff;
                    closestSum = sum;
                }
            }
        }
        
        return closestSum;
    }
}

File: Scala/3sum-closest.scala
object Solution {
    def threeSumClosest(nums: Array[Int], target: Int): Int = {
        scala.util.Sorting.quickSort(nums)
        var closestSum:Int = 0
        var closestDiff:Int = Integer.MAX_VALUE
        for(i <- 0 to nums.length-1 ){
            
            var j = i+1
            var k = nums.length-1
            
            while(j<k){
                val vsum = nums(i) + nums(j) + nums(k)
                var diff = vsum - target
                
                if(diff > 0){
                    k -= 1
                }else{
                    diff = -diff
                    j += 1
                }
                
                if(diff < closestDiff){
                    closestDiff = diff
                    closestSum = vsum
                }
                
            }
        }
        
        closestSum
    }
}
